monai/visualize/gradient_based.py

```diff
@@ -0,0 +1,163 @@
+# Copyright (c) MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from contextlib import contextmanager
+from functools import partial
+from typing import Callable
+
+import torch
+
+from monai.utils.module import optional_import
+
+trange, has_trange = optional_import("tqdm", name="trange")
+
+
+__all__ = ["VanillaGrad", "SmoothGrad", "GuidedBackpropGrad", "GuidedBackpropSmoothGrad"]
+
+
+def replace_module(parent: torch.nn.Module, name: str, new_module: torch.nn.Module) -> None:
+    idx = name.find(".")
+    if idx == -1:
+        setattr(parent, name, new_module)
+    else:
+        parent = getattr(parent, name[:idx])
+        name = name[idx + 1 :]
+        replace_module(parent, name, new_module)
+
+
+class _AutoGradReLU(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, x):
+        pos_mask = (x > 0).type_as(x)
+        output = torch.mul(x, pos_mask)
+        ctx.save_for_backward(x, output)
+        return output
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        x, _ = ctx.saved_tensors
+        pos_mask_1 = (x > 0).type_as(grad_output)
+        pos_mask_2 = (grad_output > 0).type_as(grad_output)
+        y = torch.mul(grad_output, pos_mask_1)
+        grad_input = torch.mul(y, pos_mask_2)
+        return grad_input
+
+
+class _GradReLU(torch.nn.Module):
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        out: torch.Tensor = _AutoGradReLU().apply(x)
+        return out
+
+    def backward(self, x: torch.Tensor) -> torch.Tensor:
+        out: torch.Tensor = _AutoGradReLU().backward(x)
+        return out
+
+
+class VanillaGrad:
+    def __init__(self, model: torch.nn.Module) -> None:
+        self.model = model
+
+    def get_grad(self, x: torch.Tensor, index: torch.Tensor | int | None) -> torch.Tensor:
```

Comment:
I think this class is doing a similar thing, do you think we can reuse or improve the existing one? https://github.com/Project-MONAI/MONAI/blob/facf5cc658e645c6cba57529a97902ebbddee2b7/monai/visualize/class_activation_maps.py#L137-L145

Reply:
There's certainly overlap, I'm not sure how to harmonise the classes together though, I struggle slightly with `ModelWithHooks`. Do you think you'd be able to have a look?

Reply:
Hi @rijobro this modification https://github.com/rijobro/MONAI/pull/4 passes the unit test, I havn't tested it with the updated notebook, I'll try to do it soon.

Reply:
thanks! Do you know `autograd.Function` well? It seems clunky that I've created both a `torch.autograd.Function` and a `torch.nn.Module`, I wondered if it was possible to do it with just one.

Reply:
Great, thanks.

Reply:
@wyli The results are fine, but I'm seeing this warning when using your changes on the cats and dogs notebook: 

> Backward hook for no_layer is not triggered; `requires_grad` of no_layer should be `True`.

Reply:
> thanks! Do you know `autograd.Function` well? It seems clunky that I've created both a `torch.autograd.Function` and a `torch.nn.Module`, I wondered if it was possible to do it with just one.

I think the current implementation works fine, the only minor issue is `torch.nn.Module` doesn't have `backward`, so the module backward is not really needed.